honeybee_energy/config.py

```diff
@@ -0,0 +1,322 @@
+            assert os.path.isdir(os.path.join(path, 'measures')), \
+                '{} lacks a "measures" folder.'.format(path)
+            if not self.mute:
+                print('Path to the energy_model_measure is set to: '
+                      '{}'.format(self._energy_model_measure_path))
+
+    @property
+    def python_package_path(self):
+        """Get the path to where this Python package is installed."""
+        return os.path.split(os.path.dirname(__file__))[0]
+
+    @property
+    def python_exe_path(self):
+        """Get the path to the Python.exe file."""
+        return sys.executable
+
+    def load_from_file(self, file_path):
+        """Set all of the the properties of this object from a config JSON file.
+        
+        Args:
+            file_path: Path to a JSON file containing the file paths. A sample of this
+                JSON is the config.json file within this library.
+        """
+        # check the default file path
+        assert os.path.isfile(str(file_path)), \
+            ValueError('No such a file as {}'.format(file_path))
+
+        # set the default paths to be all blank / defauktes
+        default_path = {
+            "energyplus_path": r'',
+            "openstudio_path": r'',
+            "template_library_folder": r'',
+            "default_simulation_folder": r'',
+            "energy_model_measure_path": r''
+        }
+
+        with open(file_path, 'rb') as cfg:
+            path = cfg.read()
+            try:
+                paths = json.loads(path)
+            except Exception as e:
+                print('Failed to load paths from {}.\n{}'.format(file_path, e))
+            else:
+                for key, p in paths.items():
+                    if not key.startswith('__') and p.strip():
+                        default_path[key] = p.strip()
+
+        # set paths for energyplus and openstudio installations
+        self.openstudio_path = default_path["openstudio_path"]
+        self.energyplus_path = default_path["energyplus_path"]
+
+        # set paths for the default_simulation_folder and the template_library_folder
+        self.default_simulation_folder = default_path["default_simulation_folder"]
+        self.template_library_folder = default_path["template_library_folder"]
+
+        # set the paths for the energy_model_measure
+        self.energy_model_measure_path = default_path["energy_model_measure_path"]
+
+    def _find_energy_model_measure_path(self):
+        """Find the energy_model_measure_path in its default location.
+        
+        This is usually next to the Python pacakges when it's installed for a plugin.
+        """
+        measure_path = os.path.join(self.python_package_path, 'energy_model_measure')
+        if not os.path.isdir(measure_path):
+            return  # No energy_model_measure is installed
+        return measure_path
+
+    @staticmethod
+    def _find_openstudio_folder():
```

Comment:
This is part of a larger conversation but I think all the execution related information like this one should be moved outside the libraries. The only case that I can imagine keeping them here is useful is for finding C# libraries if we decide to use them in the code.

I understand with the current design we will need them for executing the simulations but as I commented before I think we should follow a PATH based approach instead of finding the installation folder. This will also help to ensure a batch / bash file will run successfully on any machine regardless of the operating system and installation folder of the computer that is used to create it.

Reply:
I agree that a PATH-based solution would be nice and it's definitely what we should do for Radiance but I don't think that a purely PATH-based approach is going to work well for EnergyPlus/OpenStudio because the installers for EnergyPlus/OpenStudio do not include any option to add the executables to the PATH. This means that we either have to have people manually add it to the PATH, which I know isn't realistic for the majority of users attending our workshops, or we try to add it to the path using a script, which is bound to run into permissions issues on some machines unless this script is a part of a signed installer.
So, in the long run, I think a purist PATH-based approach is going to mean it works on fewer machines rather than more machines. But I would definitely be in favor of  an approach that tries to check the PATH first and then, if it doesn't find an installation there, it looks in the usual default location. The PATH-based approach will be useful on our cloud service or for those users who might buy our installer but I think we're easily going to loose the majority of the users not going this route unless we include a backup way to find the installation.
Let me know if you agree and I will implement the dual solution.

Reply:
That can work.

There will not be permission issues. You can pass the PATH when starting a subprocess. User doesn't need to know about it.

Reply:
I agree that there won't be permission issues when the user is an admin or they are using an executable installer that they had to get admin rights to run. But, if there were no permissions issues in any case, you wouldn't have people asking [questions like this](https://serverfault.com/questions/33681/how-can-i-modify-a-user-s-path-environment-variable-without-logging-out) or there wouldn't be a little shield icon next to the menu that you use to edit the PATH manually on Windows:
![image](https://user-images.githubusercontent.com/5567574/71326330-054c0900-24ae-11ea-9ef0-69bdbe352c8a.png)

Reply:
More proof that you need admin privileges to edit the PATH:
http://www.kscodes.com/misc/how-to-set-path-in-windows-without-admin-rights/

Reply:
You are confusing system variables and environmental variables for a subprocess. Those two are different. You can pass the PATH when **starting a subprocess** using the `env` argument. See the documentation here: https://docs.python.org/2/library/subprocess.html#popen-constructor

This is similar to how we set-up the PATH in every bat file. That does not need admin privileges.

Reply:
Ok, you will see that I implemented it in a way that, if there's no specification of the installation path in the config.json, it will check to see if the executable file exists in the PATH. Then, if it doesn't find it there, it will search in the default installation location and attempt to use the most recent version of EnergyPlus/OpenStudio it finds.

This means that we can eventually do some stuff with the PATH down the line when we figure out how we want to get around the admin access issue. But it will work now, even for people who don't have admin access on their machine.